src/starterbot/BaseManager.cpp

```diff
@@ -0,0 +1,156 @@
+        if (!(unit->getType().isMineralField() || unit->getType() == BWAPI::UnitTypes::Resource_Vespene_Geyser)) {
+            continue;
+        }
+
+        bw::TilePosition resourcePosition = unit->getTilePosition();
+        bool placed = false;
+
+        for (bw::Unitset& cluster : resourceClusters) {
+
+            // If this resource is close enough to an existing cluster, add it.
+            if (resourcePosition.getApproxDistance(bw::TilePosition(cluster.getPosition())) < RESOURCE_RADIUS) {
+                cluster.insert(unit);
+                placed = true;
+                break;
+            }
+        }
+
+        // Otherwise, create a new cluster.
+        if (!placed) {
+            bw::Unitset newCluster;
+            newCluster.insert(unit);
+            resourceClusters.push_back(newCluster);
+        }
+    }
+
+    // For every resource cluster found, create a new Base.
+    for (auto& cluster : resourceClusters) {
+        bw::Position centroid = cluster.getPosition();
+        std::vector<bw::Unit> minerals;
+        std::vector<bw::Unit> geysers;
+
+        // Add each resource to the appropriate list
+        for (auto& resource : cluster) {
+            if (resource->getType().isMineralField()) {
+                minerals.push_back(resource);
+            }
+            if (resource->getType() == BWAPI::UnitTypes::Resource_Vespene_Geyser) {
+                geysers.push_back(resource);
+            }
+        }
+
+        // The set of buildings and enemyCount initalize as empty and as zero, respectively
+        m_bases.push_back({ centroid, minerals, geysers, {}, 0 });
+    }
+}
+
+// Each frame, check for buildings that have been created or destroyed. Update the Base that they're
+// closest to accordingly.
+void BaseManager::onFrame() {
+
+    bw::Unitset shadowBuildings = m_unitManager.shadowUnits(
+        bw::Filter::IsBuilding &&                                   // get enemy buildings
+        !bw::Filter::IsSpecialBuilding && !bw::Filter::IsNeutral    // ignore resources and pre-exisiting buildings
+    );
+
+    // Check for new buildings.
+    for (bw::Unit building : shadowBuildings) {
+        if (m_buildings.find(building) == m_buildings.end()) {
+            m_buildings.insert(building);
+
+            // If there's a new building, find the base it now belongs to.
+            Base& closestBase = getClosestBase(building);
+
+            // Add the building to this base
+            closestBase.buildings.push_back(building);
+
+            // If the building belongs to the enemy, we increment the base's enemy count
+            if (building->getPlayer() == g_game->enemy()) {
+                closestBase.enemyCount += 1;
+            }
+        }
+    }
+
+    // Check for destroyed buildings.
+    for (bw::Unit building : m_buildings) {
+        if (shadowBuildings.find(building) == shadowBuildings.end()) {
+            m_buildings.erase(building);
```

Comment:
When iterating over a set in C++, you cannot remove the element you're currently iterating over because this triggers something known as iterator invalidation. This can cause a crash (if you're in debug mode, this should automatically be caught).

See, for instance, https://github.com/v-rob/AutoPilot/blob/main/src/starterbot/UnitManager.cpp#L157-L161 as to how you can handle this without running into iterator invalidation.